부산대 BE 이시웅 step1

silverttthin opened the pull request:
지난 단계 미션에서 정규표현식, 글자수 등의 기능은 미리구현해 놓았기에 이번 프로젝트에선 다음 2가지를 중점적으로 구현했습니다.
1. 서비스 레이어에서 '카카오' 키워드 포함 검증 로직(하단에 이 부분을 구현할 때 고민했던 부분들을 기록했습니다)
- 오늘의 고민
    - 상품 엔티티에 허가속성을 추가하고 어드민 관련 api를 만들어 담당 md 협의 하 ‘카카오’ 문구 사용 가능 여부를 결정할까?
        - 일단 response에 담당자 연락 안내하기까지만 구현하기로 결정  
        
        - 너무 단순한 하나의 요구사항 때문에 구현이 오버하게 심화될 것 같은 느낌
        - md 허가는 결국 일대일이니 걍 허가직후 어드민에서 직접 쿼리를 날리는 것으로도 요구사항 충족 되긴 할 듯?
    - 상품명 속성의 ‘카카오’ 문구 포함 여부 검증을 dto랑 서비스레이어 어디에서 할까?
        - dto에 이미 상품명 다른 검증 로직이 들어있음. 검증 책임을 줄거면 한쪽에 몰리는게 깔끔해보이는 것 같기도..?
        - 근데 저 요구사항이 단순 입력값 유효성 검증에 속할까 아니면 좀더 상세한 비즈니스 규칙에 속할까
        
    - 고민 끝에 서비스 레이어에서 하기로 결정!
        - 단순한 입력값 검증이 아닌 비즈니스 정책에 가까운 것 같음.
        - 전자는 dto에서, 후자는 서비스레이어에서 하는게 맞는 것 같음

2. 전역 핸들러 구현
    - 400번대와 500번 에러를 처리하는 전역 핸들러를 구현해 클라이언트에서 어떤 에러인지 코드와 디버깅 메시지를 통해 확인할 수 있게 했습니다.

silverttthin committed:
fix: 서비스 빈이 인터페이스가 아닌 구현체에 직접 의존하던 잘못된 부분 수정
- 하위 레이어에 의존하지 않고 유연하게 쓰려고 인터페이스 만들고 구현체 만들어둔건데 구현체에 직접 의존해버리면;

silverttthin committed:
refactor: 레이어별 프로젝트 구조 세팅

silverttthin committed:
feat: 서비스 레이어에 카카오 문구 검증로직 추가

silverttthin committed:
feat: 공용 예외처리 핸들러 구현

silverttthin committed:
feat: 수정 서비스 함수에 누락된 카카오 예외처리 로직 추가

silverttthin committed:
feat: 수정 서비스 함수에 누락된 카카오 예외처리 로직 추가

silverttthin commented:
마지막 커밋은 amend를 깜빡하고 중복 커밋메시지가 돼버렸습니다;

silverttthin committed:
feat: 400번대 에러를 분기
- valid의 경우 깔끔한 디버깅 메시지를 위한 후처리가 있기 때문에 MethodArgumentNotValid 예외는 분기

mkom76 commented on src/main/java/gift/product/controller/AdminController.java:
```diff
@@ -1,6 +1,7 @@
-package gift.product;
+package gift.product.controller;
```
패키지 구조 나누신거 좋습니다!👍👍

mkom76 commented on src/main/java/gift/product/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,26 @@
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+
+// 400번대 모든 에러와 500번 에러에 대해 상태 코드와 디버깅 메시지을 전달하는 전역 핸들러
```
클린코드 관점에서 주석은 제거해주는게 좋습니다~ 대신 클래스명, 메서드명 등으로 명확하게 나타내는걸 지향하고 있어요.
네이밍에 신경쓰시는 습관을 가지는 것도 좋습니다!!

mkom76 commented on src/main/java/gift/product/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,26 @@
+
+	public record ErrorResponse(int code, String message) {}
+
+	@ExceptionHandler({ MethodArgumentNotValidException.class, IllegalArgumentException.class })
```
전역 핸들러 구현👍👍👍

mkom76 commented on src/main/java/gift/product/exception/GlobalExceptionHandler.java:
```diff
@@ -8,19 +8,34 @@
+
+	// 카카오 키워드 포함 예외
+	@ExceptionHandler(IllegalArgumentException.class)
+	public ResponseEntity<ErrorResponse> handleIllegalArgumentException(IllegalArgumentException e) {
```
위의 메소드와 분리한 이유를 알 수 있을까요? 메세지 구조가 어떻게 다른건가요?🧐

silverttthin replied:
@valid는 에러메시지를 콘솔에 찍어보면 다음처럼 나옵니다.
```
Field error in object 'itemRequest' on field 'price': rejected value [-1]; codes [Min.itemRequest.price,Min.price,Min.java.lang.Integer,Min]; arguments [org.springframework.context.support.DefaultMessageSourceResolvable: codes [itemRequest.price,price]; arguments []; default message [price],0]; default message [가격은 음수일 수 없습니다.]
```
처음엔 이걸 그냥 응답 json의 메시지 속성에 담았는데 너무 가시성이 별로라 직접 지정한 디버깅 메시지만 추출하는 로직을 따로 적용하기 위해 메소드를 분리했습니다

mkom76 commented on src/main/java/gift/product/service/ItemService.java:
```diff
@@ -59,6 +59,9 @@ public GetItemResponse getItem(Long itemId) {
 
 	public GetItemResponse updateItem(Long itemId, ItemRequest req) {
 
+		if (req.name().contains("카카오"))
```
DTO에 두지 않는 이유는 저도 동의합니다👍
다만 혹시 이 코드를 Item 클래스에 두는 건 어떨까요? 유효성 검사 메소드를 두고, 생성자, 업데이트(추가생성) 메소드에서 호출하는 방식으로요. 유효성 검사 로직이 도메인에 있으면 좀 더 책임이 분리된 서비스를 만들 수 있을 거에요~

silverttthin replied:
좋은 개선인것 같습니다! 서비스의 여러 함수에서 동일한 기능이 중복된다면 그 기능에 관련된 엔티티 내부에 static 메서드로 만들어두면 깔끔할 것 같습니다. 다만 그 기능이 디비와 연관돼있거나 다른 도메인과 연관된 로직이라면 어쩔수 없이 서비스에 남겨놔야 할 것 같습니다. 좋은 패턴을 학습했습니다 바로 반영하겠습니다.

mkom76 reviewed (changes requested):
요구사항 잘 작성해주셨습니다👍👍👍 고생하셨어요.
테스트 코드를 살펴보니 Repository에 대한 단위테스트를 작성해주신 것 같아요. E2E 테스트를 배웠다고 들었는데 한번 구현해보시는 건 어떨까요?😊
추가 코멘트도 몇개 드렸는데 확인 부탁드립니다~!!

silverttthin committed:
comments: 주석 삭제 및 메서드 네이밍 수정
- illegalArgument 예외는 카카오 키워드 말고도 종종 쓰일 것 같아 네이밍을 따로 카카오 특화로 변경하진 않음

silverttthin committed:
refactor: 카카오 키워드 검사를 엔티티 정적 메소드로 만들어 이를 사용

silverttthin committed:
fix: 404 에러에 대한 적절한 처리 구현
- 서비스에서 runtime예외를 사용하던 존재성 검증을 nosuchelement 예외로 수정
- 핸들러에도 추가
- jdbctemplate.queryForObject 자체가 다른 예외를 먼저 터트려서 적용이 되지 않아 try catch로 감싸 그냥 null을 리턴하는 것으로 변경

silverttthin committed:
feat: e2e 테스트 구현

silverttthin commented:
e2e 테스트에 대해서도 공부해보고 리뷰어님의 피드백 반영하여 커밋했습니다. 
- e2e는 테스트 단위 중 애플리케이션의 큰 흐름, 즉 가장 큰 단위 테스트 -> 상품 crud를 하나의 테스트 메소드에 넣었고 예외 테스트는 다른 테스트 메서드로 분류했습니다.
- 중복되는 서비스 로직은 관련된 엔티티(여기선 Item)에 정적 메소드로 만들었습니다.
- 피드백 외에 추가로 존재하지 않는 상품 시 발생시키는 예외가 잘 작동하지 않는 것을 확인해 수정했습니다. 이에 대해서는 해당 커밋 메시지에 상세히 기록해놨습니다!

한가지 여쭤보고 싶은게 있었습니다.
1. 정적 메소드 관련해 구글링해보다가 도메인 클래스내에 객체를 생성하는 것 자체도 도메인 메서드화 한 팩토리 메서드라는 것을 알게 됐습니다.
2. 이렇게 되면 제가 이번에 만든 검증로직 메서드를 private화 하고 public한 팩토리 메서드 내에 삽입시켜 항상 검증된 객체 생성 로직을 갖게 하고 서비스에서도 객체 생성이 중복되니 나쁘지 않겠다 싶었습니다.
3. 하지만 제 서비스엔 살짝 적용하기 애매한 구석이 있어서 **_지금 당장하기엔 필요에 의한게 아닌 것 같단 느낌_**이 들어서 보류했습니다.
    - 실제 객체 생성 중복이 아주 많진 않음 (지금 코드 규모가 작은 편이라..)
    - 그리고 객체 생성 시 id 속성이 있는 생성자, 없는 생성자가 2개 있고 이것들이 또 따로 쓰이고 있기 때문에 조금 애매할것 같다고 생각했습니다만.........
    
글을 적으면서 곱씹어보니 id 속성이 없는 생성자를 없애버리고 그걸 사용하는 곳에서 그냥 id=null인 인스턴스를 만들어 넘겨줘도 어차피 그 로직은 id를 활용하지 않으므로 상관 없겠단 생각이 들었습니다. 객체를 생성하는 로직을 담은 팩토리 메서드를 만들 수 있겠다라고 방금 결론을 내렸습니다.

하지만 위에서 서술했듯 지금 당장의 필요에 의한게 아닌 것 같단 느낌이 조금 마음에 걸리는 것 같습니다.

mkom76 commented on src/main/java/gift/product/exception/GlobalExceptionHandler.java:
```diff
@@ -13,25 +13,24 @@ public class GlobalExceptionHandler {
+
 	@ExceptionHandler(MethodArgumentNotValidException.class)
-	public ResponseEntity<ErrorResponse> handleMethodArgumentNotValidException(MethodArgumentNotValidException e) {
+	public ResponseEntity<ErrorResponse> handleMethodArgumentNotValidExceptionForValid(MethodArgumentNotValidException e) {
```
좀 더 명확한 네이밍 좋습니다👍

mkom76 commented on src/main/java/gift/product/entity/Item.java:
```diff
@@ -24,6 +24,11 @@ public Item(Long id, String name, Integer price, String imageUrl) {
 		this.imageUrl = imageUrl;
 	}
 
+	public static void validateKakaoKeyword(String name){
```
도메인 클래스의 책임으로 잘 넘겨주셨네요.😊😊
질문에 대한 답변을 여기서 드릴게요.
'정적 팩토리 메소드를 사용하기에 프로젝트 규모가 작다' -> 맞는 말씁입니다만, 저희는 추후 굉장히 커질 대규모 서비스를 염두에 두고 공부하고 있기 때문에 이번에 경험해보시는 것도 좋다는 의견 드려요😊

두번째에 대해서는 많이 사용하는 솔루션을 제시해드릴게요.

```
	public Item(Long id, String name, Integer price, String imageUrl) {
                // validation logic
		this.id = id;
		this.name = name;
		this.price = price;
		this.imageUrl = imageUrl;
	}
    public Item(String name, Integer price, String imageUrl) {
	        return this(null, name, price, imageUrl);
	}
```

이 코드 한번 참고해주세요.😊 답변이 되었으면 좋겠네요.

위 코드를 참고하여 내부 private 메소드로 선언하여 생성자에 추가하는 거는 어떨까요? update 메소드도 추가하여 안에서 호출하는 방법도 고려해보세요.😊

mkom76 reviewed (approved):
리뷰 반영하시느라 고생하셨습니다. 테스트도 잘 작성해주셨네요👍👍
몇가지 코멘트 드렸는데 이는 다음 단계에서 함께 반영해보세요😊
다음 단계도 화이팅입니다!💪💪
